src/store.ts

```diff
@@ -351,35 +401,239 @@ export default class Store {
+      contestOffset,
+      contestOffset + layout.contests.length
+    )
+
+    assert.equal(contestDefinitions.length, layout.contests.length)
+
+    const contests: Contest[] = contestDefinitions.map(
+      (contestDefinition, contestIndex) => {
+        const contestLayout = layout.contests[contestIndex]
+        const contestMarksOffset = layout.contests
+          .slice(0, contestIndex)
+          .reduce((sum, contest) => sum + contest.options.length, 0)
+        const contestMarks = marks.marks.slice(
+          contestMarksOffset,
+          contestMarksOffset + contestLayout.options.length
+        )
+        const options: ContestOption[] =
+          contestDefinition.type === 'candidate'
+            ? contestMarks.map<CandidateContestOption>((mark, markIndex) => {
+                assert.notEqual(mark.type, 'stray')
+
+                const candidate: Candidate | undefined =
+                  contestDefinition.candidates[markIndex]
+
+                return {
+                  type: 'candidate',
+                  id: candidate?.id ?? '__write-in',
+                  name: candidate?.name ?? 'Write-In',
+                  bounds: contestLayout.options[markIndex].bounds,
+                }
+              })
+            : contestMarks.map<YesNoContestOption>((mark, markIndex) => {
+                assert.notEqual(mark.type, 'stray')
+
+                return {
+                  type: 'yesno',
+                  id: markIndex === 0 ? 'yes' : 'no',
+                  name: markIndex === 0 ? 'yes' : 'no',
+                  bounds: contestLayout.options[markIndex].bounds,
+                }
+              })
+
+        return {
+          id: contestDefinition.id,
+          title: contestDefinition.title,
+          bounds: contestLayout.bounds,
+          options,
+        }
+      }
+    )
+
+    debug('overlaying ballot adjudication: %O', adjudication)
+
+    const overlay = marks.marks.reduce<MarksByContestId>(
+      (marks, mark) =>
+        mark.type === 'stray'
+          ? marks
+          : {
+              ...marks,
+              [mark.contest.id]: {
+                ...marks[mark.contest.id],
+                [typeof mark.option === 'string'
+                  ? mark.option
+                  : mark.option.id]:
+                  adjudication?.[mark.contest.id]?.[
+                    typeof mark.option === 'string'
+                      ? mark.option
+                      : mark.option.id
+                  ] ??
+                  isMarked(mark) ??
+                  false,
+              },
+            },
+      {}
+    )
+
+    debug('overlay results: %O', overlay)
+
     return {
-      id: row.id,
-      filename: row.filename,
-      marks: row.marksJSON ? JSON.parse(row.marksJSON) : undefined,
-      cvr: row.cvrJSON ? JSON.parse(row.cvrJSON) : undefined,
+      ballot,
+      marks: overlay,
+      contests,
+    }
+  }
+
+  public async adjudicateBallot(
```

Comment:
pedantic nit, `saveBallotAdjudication` or some other naming where the name unambiguously indicates a change in what's stored.

Reply:
Done.